### DISCUSSION TITLE
What is the correct method to implement a PID controller with a Process Model?

### DISCUSSION THREAD

**Discussion by TheWicklowWolf** [python-control/python-control#845] [Q&A] [answered]
I would like to simulate a system response from FOPDT model with a PID controller. 
This is what I have so far:

- How do I limit the output of the PID to 0-100?
- How can I include anti-windup on the Integral (based on 0-100 limits)?
- Is there a way to set the Initial state for Integral specifically?

Any suggestions are appreciated.

```
import control
import numpy as np
import matplotlib.pyplot as plt

gain=2
tc=20
dt=5
bias=13.5

Kp = 1
Ki = 0.05
Kd = 1.5

# First-order plus deadtime system
num=np.array([gain])
den=np.array([tc,1])
#Add dead time
n_pade = 10
(num_pade,den_pade) = control.pade(dt,n_pade)
H_pade=control.tf(num_pade,den_pade)
H_noDelay = control.tf(num,den)
H = control.series(H_pade,H_noDelay)
# PID controller
C = control.tf([Kd, Kp, Ki], [1, 0])
# First-order plus deadtime system and PID Controller
G = control.feedback(H*C, 1)
t = np.linspace(0, 100, 100)
t, pv = control.step_response(G, t)

sp=np.ones(len(pv))

# Plot the results
plt.plot(t, sp, label='SP')
plt.plot(t, pv, label='PV')
plt.legend()
plt.show()
```

![image](https://user-images.githubusercontent.com/111055425/211327532-b837b1d4-040b-40b6-be96-e7b4d879e73c.png)

**Comment by murrayrm** [comment, marked as answer]
To implement saturation nonlinearities and anti-windup compensation, you'll have to create an input/output (I/O) system description.  An example of a PID controller with anti-windup protection can be found in one of the examples in the documentation: [PI controller with antiwindup protection](https://python-control.readthedocs.io/en/latest/cruise.html#PI-controller-with-antiwindup-protection).  Note that the saturation nonlinearity is in the process dynamics (look for the `clip` call [here](https://python-control.readthedocs.io/en/latest/cruise.html#Process-Model), but you could also clip the output of the PID controller.

**Reply by TheWicklowWolf** [reply]
Thanks for that, a little of parsing required. Are there any quick-start/basic examples?
I was hoping someone has a stripped out version to get started with.
Thanks again.
